Make a distinction between hand-offs for rebalance and region shutdown

octonato opened the pull request:
References #29549

Not sure if I'm taking the right path here. Also some ideas on how to test it will be welcome.

renatocaval commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -719,7 +721,11 @@ abstract class ShardCoordinator(
-              continueRebalance(shards.toSet)
+
+              // only for rebalance if we have at least one active region
+              if (calculateActiveRegions() > 1)
+                continueRebalance(shards.toSet)
+
```
I initially thought about simply check if there is more than one region, but then thought about the following scenario:

* Cluster with two regions
* both regions start to shutdown
* `GracefulShutdownReq` from first region is received, coordinator runs a rebalance (we can't avoid this one)
* second region sends a `GracefulShutdownReq`, first region is not yet gone, but it makes no sense to run a rebalance. So we skip it.

renatocaval commented:
I think my implementation here is completely missing the point. I tried to write a multi-jvm tests and got some new insight. I will re-work this tomorrow.

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -719,7 +721,11 @@ abstract class ShardCoordinator(
+
+              // only for rebalance if we have at least one active region
+              if (calculateActiveRegions().size > 1)
+                continueRebalance(shards.toSet)
```
This seems right to me.
One question is if we should have similar condition in `case RebalanceTick`? It's currently looking at `state.regions`, which is the stored state. Maybe it should also avoid rebalance in same way?

renatocaval commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -556,6 +556,8 @@ abstract class ShardCoordinator(
   var aliveRegions = Set.empty[ActorRef]
   var regionTerminationInProgress = Set.empty[ActorRef]
 
+  private def calculateActiveRegions() = (state.regions -- gracefulShutdownInProgress) -- regionTerminationInProgress
```
I realised this is wrong. We should not exclude regions that are in progress of shutdown. It's possible that a shard in such a region fails to shutdown. As a result the region will still around. 

If we exclude them here, we maybe have regions around with invalid shard caches.

renatocaval commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -757,11 +786,38 @@ abstract class ShardCoordinator(
         } else {
+          log.warning("Rebalance shard [{}] didn't complete within [{}].", shard, handOffTimeout.pretty)
+          clearRebalanceInProgress(shard)
+        }
```
The handling of `RebalanceDone` doesn't need to take the graceful shutdown into account. So, the if and else logic is simplified.

renatocaval commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -757,11 +786,38 @@ abstract class ShardCoordinator(
+          clearRebalanceInProgress(shard)
+        }
+
+      case ShardShutdownDone(shard, ok) =>
+        rebalanceWorkers -= sender()
+
+        if (ok) {
+          log.debug("Shard [{}] shutdown completed successfully.", shard)
+          if (state.shards.contains(shard)) {
+            update(ShardHomeDeallocated(shard)) { evt =>
+              log.debug("Shard [{}] deallocated after after region shutdown", shard)
+              state = state.updated(evt)
+              clearRebalanceInProgress(shard)
+              allocateShardHomesForRememberEntities()
+              self.tell(GetShardHome(shard), ignoreRef)
+            }
+          } else {
+            clearRebalanceInProgress(shard)
+          }
+        } else {
+          log.warning("Shard [{}] shutdown didn't complete within [{}].", shard, handOffTimeout.pretty)
+          // TODO: we should take that into consideration for future rebalances
+          // the region is trying to shutdown but hangs because of a shard,
+          // we may want to not allocate new shards on this region
+
+          // as a shutdown failed we need to put region back on the shelf
+          // graceful shutdown will be retried
+          gracefulShutdownInProgress -= state.shards(shard)
```
here we know that we are dealing with a graceful shutdown, so in case of failure we put the region back to 'active'.

patriknw replied:
I'm not sure it is better to have two separate cases for `ShardShutdownDone` and `RebalanceDone`. They should do pretty much the same aside from some logging. If it's important to adjust the logging here also it's possible to see if it's a rebalance or shutdown by looking at `gracefulShutdownInProgress.contains(state.shards(shard))`

renatocaval commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -1035,11 +1092,35 @@ abstract class ShardCoordinator(
+    }
+  }
 
+  def shutdownShards(shards: Set[ShardId]): Unit = {
+    if ((log: BusLogging).isInfoEnabled && (shards.nonEmpty)) {
+      log.info("Starting shutting down shards [{}].", shards.mkString(","))
+    }
+    shards.foreach { shard =>
+      if (!rebalanceInProgress.contains(shard)) {
+        state.shards.get(shard) match {
+          case Some(shuttingDownFromRegion) =>
+            rebalanceInProgress = rebalanceInProgress.updated(shard, Set.empty)
+            log.debug("Shutting down shard [{}] from [{}]", shard, shuttingDownFromRegion)
+            rebalanceWorkers += context.actorOf(
+              shutdownWorkerProps(
+                shard,
+                shuttingDownFromRegion,
+                handOffTimeout,
+                state.regions.keySet.union(state.regionProxies)
+              ).withDispatcher(context.props.dispatcher))
+          case None =>
+            log.debug("Shutting down of non-existing shard [{}] is ignored", shard)
+        }
```
Nearly the same as `continueRebalance`, but for region shutdown.

octonato reviewed (commented):
Did a U-turn and, as discussed today, included different paths for rebalance and region shutdown.

I have some other stuff to add here, but will send different PRs on top of this one because not yet sure they make sense.

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -525,7 +532,14 @@ object ShardCoordinator {
 
-    log.debug("Rebalance [{}] from region [{}]", shard, regions)
+    if (isRebalance)
+      log.debug("Rebalance [{}] from region [{}]", shard, regions)
```
In https://github.com/akka/akka/pull/29590 I tried to get away from logging full Set of regions, and instead use the size. Maybe this PR isn't rebased?

Could you also fix my parameter order mistake? https://github.com/akka/akka/pull/29590/files#r487298916

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -537,7 +551,11 @@ object ShardCoordinator {
+        if (isRebalance)
+          log.debug("Rebalance of [{}]  from [{}] timed out", shard, from)
+        else
+          log.debug("Hand-off of [{}] shard from [{}] timed out", shard, from)
```
maybe start this log message in same way as above? "Shutting down shard..."

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -757,11 +786,38 @@ abstract class ShardCoordinator(
+          log.warning("Shard [{}] shutdown didn't complete within [{}].", shard, handOffTimeout.pretty)
+          // TODO: we should take that into consideration for future rebalances
+          // the region is trying to shutdown but hangs because of a shard,
+          // we may want to not allocate new shards on this region
```
I think that is handled (at least after a while) because we avoid allocating to leaving members. It's possible that we haven't seen the leaving status here yet, but I'm not sure it's worth keeping track of another "in progress" thing.

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -1039,7 +1096,29 @@ abstract class ShardCoordinator(
+    }
+  }
 
+  def shutdownShards(shards: Set[ShardId]): Unit = {
```
The only difference between continueRebalance and shutdownShards is again the logging, unless I missed something. I don't think we should duplicate the code just because of that. Better to pass `rebalance: Boolean` flag as a parameter.

octonato committed:
distinct rebalance from region shutdown

renatocaval commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -1039,7 +1096,29 @@ abstract class ShardCoordinator(
+  }
 
+  def shutdownShards(shards: Set[ShardId]): Unit = {
+    if ((log: BusLogging).isInfoEnabled && (shards.nonEmpty)) {
```
seems to be, I got this error after removing it:

```scala
ambiguous reference to overloaded definition,
[error] both method isInfoEnabled in class MarkerLoggingAdapter of type (marker: akka.event.LogMarker)Boolean
[error] and  method isInfoEnabled in class BusLogging of type => Boolean
[error] match expected type ?
```

renatocaval replied:
@patriknw, you had previously made a comment about the duplication in `continueRebalance` and `shutdownShards`. 

I re-worked this method and I think it's worth separating them.

When shutting down, we have already the region at hand, so we don't need to fetch it and have match `Some/None`. 

I did move the common parts to `startShardRebalanceIfNeeded`. This method now takes care of all the shard rebalance including checking if it's needs to create an worker or not.

renatocaval commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -743,26 +774,35 @@ abstract class ShardCoordinator(
+        if (ok) {
+          log.debug("Shard [{}] {} completed successfully.", shard, logContext)
+
+          // The shard could have been removed by ShardRegionTerminated
+          if (state.shards.contains(shard)) {
             update(ShardHomeDeallocated(shard)) { evt =>
-              log.debug("Shard [{}] deallocated after rebalance", shard)
+              log.debug("Shard [{}] deallocated after {}", shard, logContext)
               state = state.updated(evt)
               clearRebalanceInProgress(shard)
               allocateShardHomesForRememberEntities()
               self.tell(GetShardHome(shard), ignoreRef)
             }
           } else {
-            // rebalance not completed, graceful shutdown will be retried
-            gracefulShutdownInProgress -= state.shards(shard)
             clearRebalanceInProgress(shard)
           }
+
         } else {
+          log.warning("Shard [{}] {} didn't complete within [{}].", shard, logContext, handOffTimeout.pretty)
+
+          // was that due to a graceful region shutdown?
+          // if so, consider the region as still alive and let it retry to gracefully shutdown later
+          if (gracefulShutdownInProgress(state.shards(shard)))
+            gracefulShutdownInProgress -= state.shards(shard)
+
```
Extra attention to this part. 

The functionality didn't change, but I re-organised the if/else by reducing the amount of branching. 

In a nutshell, 

if `OK` and shard is in state, update the state, otherwise clear rebalance
if `NOK`, remove region from graceful shutdown if needed and clear rebalance

This is essentially what was happening before, but with more nesting and checks.

renatocaval commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -743,26 +774,35 @@ abstract class ShardCoordinator(
-        if (state.shards.contains(shard)) {
-          if (ok) {
+
+        val logContext =
+          if (gracefulShutdownInProgress(state.shards(shard))) "shutdown"
+          else "rebalance"
```
I dislike this and I see a potential confusion logging. We can have a general rebalance followed by a region shutdown. Then, we get the `RebalanceDone` and we log that the shard is dealloacted because of a shutdown. 

Not big issue I think. At the end, there is a shutdown happening anyway. So, we are not really lying. ;-)

renatocaval replied:
I would prefer to not have `logContext` and simply say: "Shard dealloacted"

WDYT?

patriknw replied:
Ok, we can use "deallocated".

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -499,7 +509,9 @@ object ShardCoordinator {
   private[akka] object RebalanceWorker {
+
     final case class ShardRegionTerminated(region: ActorRef)
+
```
spacy (not sure it's better)

renatocaval replied:
I don't remember to have done that. There are other formatting above. 

Anyway, will revert and build again to see what we get.

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -514,18 +526,27 @@ object ShardCoordinator {
 
-    log.debug("Rebalance [{}] from [{}] regions", regions.size, shard)
+    if (isRebalance)
+      log.debug("Rebalance [{}] from [{}] regions", shard, regions.size)
```
thanks

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -743,26 +774,35 @@ abstract class ShardCoordinator(
+
+          // was that due to a graceful region shutdown?
+          // if so, consider the region as still alive and let it retry to gracefully shutdown later
+          if (gracefulShutdownInProgress(state.shards(shard)))
```
That `if` is not needed. If it's not in the Set it still ok to `-=`.

patriknw commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -1047,25 +1109,28 @@ abstract class ShardCoordinator(
+      log.info("Starting shutting down shards [{}] due to region shutting down.", shards.mkString(","))
+    }
+    shards.foreach { shard =>
+      log.debug("Shutting down shard [{}] from [{}]", shard, shuttingDownRegion)
```
isn't that logging redundant with above `log.info("Starting shutting` ?

patriknw reviewed (commented):
this is good, just a few small things...

renatocaval commented:
@patriknw / @johanandren, I applied the code review suggestions. This is ready for another round.

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -534,10 +546,16 @@ object ShardCoordinator {
       case ReceiveTimeout =>
-        log.debug("Rebalance of [{}]  from [{}] timed out", shard, from)
+        if (isRebalance)
+          log.debug("Rebalance of [{}]  from [{}] timed out", shard, from)
```
double space in log string

renatocaval replied:
```suggestion
          log.debug("Rebalance of [{}] from [{}] timed out", shard, from)
```

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardRegion.scala:
```diff
@@ -614,7 +614,7 @@ private[akka] class ShardRegion(
    * the likely locations of the coordinator.
    */
-  def coordinatorSelection: List[ActorSelection] = {
+  def coordinatorSelection(): List[ActorSelection] = {
```
This is side effect free so without parenthesises was fine?

renatocaval replied:
Yes, I agree. I added it because it was being used on a method twice ([here](https://github.com/akka/akka/pull/29579/files#diff-9d06486133d0565a63482bccb91579c0L1173-L1174)).

Probably because it looked like a `val`. If it had a `()`, we would probably think twice before calling it like that.

johanandren replied:
Not that it matters that much but, we don't usually use `()` to signal anything but side effects, so a bit weird using it for something else here.

johanandren commented on akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala:
```diff
@@ -743,26 +765,30 @@ abstract class ShardCoordinator(
+
+          // was that due to a graceful region shutdown?
+          // if so, consider the region as still alive and let it retry to gracefully shutdown later
+          gracefulShutdownInProgress -= state.shards(shard)
```
I don't get this, if the region did shutdown gracefully there will be nobody retrying later?

Also, no couldn't the region have terminated, so that shard no longer is in `state.shards` and causing this to throw `java.util.NoSuchElementException`?

renatocaval replied:
You are right about the `NoSuchElementException`. Needs a guard.

About the graceful shutdown, this was already like that (see previous code).

This code only executes if `RebalanceDone(_, ok = false)`. It will be `false` if we get a timeout when waiting for the `BeginHandOffAck`.

If the timeout happens because of a delay and it was a graceful shutdown, then the region will send, soon or later, another `GracefulShutdownReq`. This is fine in that case. 

If the timeout happens because the node crashed, removing it from the set of `gracefulShutdownInProgress` is not an issue. Soon or later, it will be removed from the list of registered shard regions. 

However, this bring us to another discussion. Something that I have being trying to wrap my head around a few times while changing that code. 

If a node ever sends a `GracefulShutdownReq`, we should not bring it back to the set of active shard regions (like we are doing here). We don't want to allocate there anymore. We do want it to clear its cache and send buffered messages as soon as possible. We want to include it on rebalance, but only to send it hand-off messages.

johanandren replied:
I also find that curious.

patriknw replied:
NoSuchElementException? trying to remove an element from a Set that doesn't exist in the Set is fine, it will be same Set

renatocaval replied:
NoSuchElement when trying to find it in the Map.

renatocaval replied:
@johanandren, I think I understand it now.

If it get stuck, it will send another request to the coordinator, but the coordinator will only let it go if not on `gracefulShutdownInProgress`. See the handling of `GracefulShutdownReq` below. 

We need that dance between the parts. 

I think that it will be more efficient if we don't put it back on the list of active regions. It said it's leaving, so it needs to go. 

The region can keep try to shutdown itself and the coordinator can stop to allocate stuff on it and when we don't need all that back and forth.

renatocaval replied:
Anyway, I'm not suggesting we should change it now, but we can review this in baby-steps. For me is very helpful to wrap my mind around it. From time to time to small improvements will be beneficial.

patriknw reviewed (approved):
LGTM

octonato commented:
Rebased with master and squashed the commits. 

@johanandren, I removed the `()` from `coordinatorSelection`. 

Apart from that, there is no other changes except some merge conflict resolution.

johanandren commented:
Make it green and I'll give you a LGTM

octonato committed:
fixing mima for a greenish life

octonato commented:
@johanandren, on its way. 

We should be more diligent with methods inside actors. We should have them all private to avoid mima warnings.
